### PR TITLE
[AIRFLOW-3741] Add extra config to Oracle hook

### PR DESCRIPTION
**Pull request by zhongjiajie**
Add extra config to Oracle hook
including encoding, mode, threaded etc

Add Oracle hook unittest

Add docs in insert_rows and
bulk_insert_rows

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3741
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Add extra config to Oracle hook

### Tests

- [x] My PR adds the following unit tests : 

test_oracle_hook.py

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by zhongjiajie** [comment]
Thanks god. pass the CI test finally.
@ashb @feng-tao @kaxil PTAL. Thanks
One question. Why the CI test failed in no reason?

**Comment by ashb** [comment]
>  One question. Why the CI test failed in no reason?

Because they are flaky and no one has yet found the time to dig in to why :(

File: airflow/hooks/oracle_hook.py [hunk 1]
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
...
+            conn_config['dsn'] = cx_Oracle.makedsn(dsn, conn.port, sid)
         elif dsn and service_name and not sid:
-            dsn = cx_Oracle.makedsn(dsn, conn.port, service_name=service_name)
-            conn = cx_Oracle.connect(conn.login, conn.password, dsn=dsn)
+            conn_config['dsn'] = cx_Oracle.makedsn(dsn, conn.port, service_name=service_name)
         else:
-            conn = cx_Oracle.connect(conn.login, conn.password, conn.host)
+            conn_config['dsn'] = conn.host
 
+        if conn.extra_dejson.get('encoding', None):
```
Comment (ashb) [inline airflow/hooks/oracle_hook.py]:
```suggestion
        if "encoding" in conn.extra_dejson:
```

And similar changes for all of the other `if get(x. None)` conditions.

File: airflow/hooks/oracle_hook.py [hunk 2]
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
...
+            conn_config['encoding'] = conn.extra_dejson.get('encoding')
+            # if `encoding` is specific but `nencoding` is not
+            # `nencoding` should use same values as `encoding` to set encoding, inspired by
+            # https://github.com/oracle/python-cx_Oracle/issues/157#issuecomment-371877993
+            if not conn.extra_dejson.get('nencoding', None):
+                conn_config['nencoding'] = conn.extra_dejson.get('encoding')
+        if conn.extra_dejson.get('nencoding', None):
+            conn_config['nencoding'] = conn.extra_dejson.get('nencoding')
+        mode = conn.extra_dejson.get('mode', None)
```
Comment (ashb) [inline airflow/hooks/oracle_hook.py]:
```suggestion
        mode = conn.extra_dejson.get('mode', '').lower()
```

And then we don't need `lower()` on all of the branches, or the extra `if mode:` on line 79

File: airflow/hooks/oracle_hook.py [hunk 3]
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
...
+            elif mode.lower() == 'sysbkp':
+                conn_config['mode'] = cx_Oracle.SYSBKP
+            elif mode.lower() == 'sysdgd':
+                conn_config['mode'] = cx_Oracle.SYSDGD
+            elif mode.lower() == 'syskmt':
+                conn_config['mode'] = cx_Oracle.SYSKMT
+            elif mode.lower() == 'sysrac':
+                conn_config['mode'] = cx_Oracle.SYSRAC
+        if conn.extra_dejson.get('threaded', None):
+            conn_config['threaded'] = True
```
Comment (ashb) [inline airflow/hooks/oracle_hook.py]:
```suggestion
            conn_config['threaded'] = conn.extra_dejson['threaded']
```

And similar for events.

File: airflow/hooks/oracle_hook.py [hunk 4]
```diff
@@ -128,13 +181,33 @@ def bulk_insert_rows(self, table, rows, target_fields=None, commit_every=5000):
...
+            Default 5000. Set greater than 0. Set 1 to insert each row in each transaction
+        :type commit_every: int
         """
+        if not rows:
+            raise ValueError("rows could not be null or None.")
         conn = self.get_conn()
         cursor = conn.cursor()
-        values = ', '.join(':%s' % i for i in range(1, len(target_fields) + 1))
-        prepared_stm = 'insert into {tablename} ({columns}) values ({values})'.format(
+        if target_fields:
```
Comment (ashb) [inline airflow/hooks/oracle_hook.py]: This change looks unrelated to the config change. what's this for?

File: tests/hooks/test_oracle_hook.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import json
+import unittest
+from datetime import datetime
+
+import cx_Oracle
```
Comment (ashb) [inline tests/hooks/test_oracle_hook.py]: Could you guard this in an try/except and decorate the whole class with a `@unittest.skipIf`?

See tests_mongo_hook for an example (but only one decorator on the whole class, not one per fn please)

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/hooks/oracle_hook.py (hunk 1 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: That great

File: airflow/hooks/oracle_hook.py (hunk 2 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: That great

File: airflow/hooks/oracle_hook.py (hunk 3 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: @ashb I do that because parameters only accept `True` or `False`, I wonder user will pass other value

File: airflow/hooks/oracle_hook.py (hunk 4 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: @ashb This change ref the sub commit message
> Fix Oracle hook bulk_insert_rows bug
when target_fields use default None
value

Old code define `target_fields=None` but when target_fields is None both 
https://github.com/apache/airflow/blob/2c7bb17435926628914571f14dad3c45bc53753b/airflow/hooks/oracle_hook.py#L136
and 
https://github.com/apache/airflow/blob/2c7bb17435926628914571f14dad3c45bc53753b/airflow/hooks/oracle_hook.py#L139
will raise Error

So this is to fix that

File: airflow/hooks/oracle_hook.py (hunk 3 above)
Comment (ashb) [reply airflow/hooks/oracle_hook.py]: I wouldn't worry about that. More confusing would be a user specifying `"threaded": false` and have it turn on!

Just document it (or point at the oracle docs) as boolean and keep it simple on our side

File: airflow/hooks/oracle_hook.py (hunk 4 above)
Comment (ashb) [reply airflow/hooks/oracle_hook.py]: Your PR is "Add extra config to Oracle hook" - that sounds like a different issue. Please could you open that as a separate PR?

File: airflow/hooks/oracle_hook.py [hunk 6]
```diff
@@ -50,20 +50,61 @@ def get_conn(self):
         as in ``{ "dsn":"some.host.address" , "service_name":"some.service.name" }``
         """
         conn = self.get_connection(self.oracle_conn_id)
+        conn_config = {
+            'user': conn.login,
+            'password': conn.password or ''
```
Comment (ashb) [inline airflow/hooks/oracle_hook.py]: Is empty string different to None? I'd probably say the `or ''` isn't needed, but not certain with Oracle.

File: airflow/hooks/oracle_hook.py (hunk 3 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: Ok, I get it

File: airflow/hooks/oracle_hook.py (hunk 4 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: Ok, I will submit new PR to do that.

File: airflow/hooks/oracle_hook.py (hunk 6 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: You're right, my fault.

File: airflow/hooks/oracle_hook.py (hunk 1 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py]: fix

**Comment by zhongjiajie** [comment]
@ashb All fix, PTAL, thx

File: airflow/hooks/oracle_hook.py [hunk 7]
```diff
@@ -48,22 +48,65 @@ def get_conn(self):
               that you are connecting to (CONNECT_DATA part of TNS)
         You can set these parameters in the extra fields of your connection
         as in ``{ "dsn":"some.host.address" , "service_name":"some.service.name" }``
+        see more param detail in
+        ``https://cx-oracle.readthedocs.io/en/latest/module.html#cx_Oracle.connect``
         """
         conn = self.get_connection(self.oracle_conn_id)
+        conn_config = {
+            'user': conn.login,
```
Comment (kaxil) [inline airflow/hooks/oracle_hook.py:57]: Maybe add a condition block to check if `login` & `password` fields are not empty and raise an error that they are required if it is empty.

**Comment by mik-laj** [comment]
We have not described how to create an Oracle connection in the documentation. Would you find time to fill in the gaps in the documentation? You should look at `docs/howto/manage-connections.rst` file. If you do not have time, I understand.

**Comment by zhongjiajie** [comment]
@mik-laj Ok.I will do today when I have time. thx

File: airflow/hooks/oracle_hook.py (hunk 7 above)
Comment (zhongjiajie) [reply airflow/hooks/oracle_hook.py:57]: @kaxil When I write code. I think empty usr or pwd would raise by cx_Oracle.connect in L109. What do you think. Should I write raise error block?

**Comment by zhongjiajie** [comment]
@mik-laj I have add documentation. @ashb @kaxil PTAL, thanks. We still have a discuss at https://github.com/apache/airflow/pull/4584#discussion_r251136895.
But I think error should raise by `cx_Oracle` itself.

**Comment by zhongjiajie** [comment]
ping @ashb and @kaxil again

File: airflow/hooks/oracle_hook.py (hunk 7 above)
Comment (ashb) [reply airflow/hooks/oracle_hook.py:57]: This isn't a regression or a change in behaviour, so I'll merge this.

**Review by ashb — APPROVED** [review]

**Comment by zhongjiajie** [comment]
@ashb Thanks for changing and reviewing. But CI test in `check docs` section, should I restart CI test or not?

**Comment by ashb** [comment]
@zhongjiajie Oh it failed with a random timeout. I've restarted it. Hopefully with my changes above it should now pass. (Hope you don't mind that I edited your PR - it seemed easier to do via the UI at the time)

**Comment by zhongjiajie** [comment]
@ashb I don't mind your edited and I thanks for what you did. I review the change, it should pass the CI test.

**Comment by zhongjiajie** [comment]
@ashb FYI, CI test pass
